### ISSUE TITLE
[RFC] Optimize Compiler Performance by Evaluating Pre-Prepacked Functions

### ISSUE THREAD

**Issue by sebmarkbage** [facebook/prepack#2280]
I had an idea for optimizing the compiler performance.

Finding the fixed point for functions called by a pure function is pretty. Even if those functions are themselves not pure because they can mutate arguments.

The first time we encounter a new function, we can first call it with abstract arguments. That can possibly simplify the function significantly by folding the module scope constants. That gives us all the effects from the function.

Then we can evaluate the effects by replacing the abstract arguments with concrete ones. Generating a specialized form of that function. Similar to the simplifier.

That way we don't have to redo all the work for functions we've already visited in a previous branch.

**Comment by NTillmann** [comment]
Technically that should work, but it could also result in a performance penalty: if the there's any significant branching in the function (or even a loop), or if the function is only ever called once then this is also pure overhead.

Maybe Prepack could maintain some kind of "hint" cache on the side to remember where it was worthwhile.

**Comment by calebmer** [comment]
I had assumed before this was roughly how Prepack worked.

Could we “clone” the abstract result and then patch it with concrete values? Instead of reinterpreting the function.

**Comment by NTillmann** [comment]
> Could we “clone” the abstract result and then patch it with concrete values? Instead of reinterpreting the function.

I don't think we have this functionality today, but it should be simple visitor over abstract values to instantiate them / their `args` and re-run the simplifier over them (of course, never mutating, but producing a new one whenever there's something to replace).

**Comment by sebmarkbage** [comment]
Note that even patching it with concrete values will likely reinterpret parts of it. E.g. any operation that havocs currently is left as residual and trigger all kinds of effects when we pass it real values.
